pkg/registration/hub/gc/controller.go

```diff
@@ -0,0 +1,176 @@
+type gcReconciler interface {
+	reconcile(ctx context.Context, clusterName string, cluster *clusterv1.ManagedCluster) (gcReconcileOp, error)
+}
+
+type GCController struct {
+	clusterLister   clusterv1listers.ManagedClusterLister
+	namespaceLister corev1listers.NamespaceLister
+	clusterPatcher  patcher.Patcher[*clusterv1.ManagedCluster, clusterv1.ManagedClusterSpec, clusterv1.ManagedClusterStatus]
+	gcReconcilers   []gcReconciler
+}
+
+// NewGCController ensures the related resources are cleaned up after cluster is deleted
+func NewGCController(
+	clusterRoleLister rbacv1listers.ClusterRoleLister,
+	clusterRoleBindingLister rbacv1listers.ClusterRoleBindingLister,
+	roleBindingLister rbacv1listers.RoleBindingLister,
+	namespaceInformer corev1informers.NamespaceInformer,
+	clusterInformer informerv1.ManagedClusterInformer,
+	manifestWorkLister worklister.ManifestWorkLister,
+	clusterClient clientset.Interface,
+	kubeClient kubernetes.Interface,
+	metadataClient metadata.Interface,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	clusterPatcher := patcher.NewPatcher[
+		*clusterv1.ManagedCluster, clusterv1.ManagedClusterSpec, clusterv1.ManagedClusterStatus](
+		clusterClient.ClusterV1().ManagedClusters())
+	controller := &GCController{
+		clusterLister:   clusterInformer.Lister(),
+		namespaceLister: namespaceInformer.Lister(),
+		clusterPatcher:  clusterPatcher,
+		gcReconcilers: []gcReconciler{
+			// currently only support to gc addons and works, need consider to add rbac to registration if gc other resources.
+			newGCResourcesController(metadataClient, []resourceFilter{
+				{addonGvr, ""}, {workGvr, getGCWorkLabelSelector()}}, eventRecorder),
+			newGCClusterRbacController(kubeClient, clusterPatcher, clusterInformer, clusterRoleLister,
+				clusterRoleBindingLister, roleBindingLister, manifestWorkLister, eventRecorder),
+			newGCClusterRoleController(clusterRoleLister, clusterInformer.Lister(), manifestWorkLister, kubeClient.RbacV1(), eventRecorder),
+		},
+	}
+
+	return factory.New().
+		WithInformersQueueKeysFunc(queue.QueueKeyByMetaName, clusterInformer.Informer(), namespaceInformer.Informer()).
+		WithSync(controller.sync).ToController("GCController", eventRecorder)
+}
+
+// gc controller is watching cluster and ns and to do these jobs:
+//  1. add a cleanup finalizer to managedCluster if the cluster is not deleting.
+//  2. clean up all rbac and resources in the cluster ns when the cluster is deleting
+//     or the cluster is deleted and cluster ns is deleting.
+//  3. clean up registration and work clusterRoles if there is no cluster and manifestWorks.
+func (r *GCController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	clusterName := controllerContext.QueueKey()
+	if clusterName == "" || clusterName == factory.DefaultQueueKey {
+		return nil
+	}
+
+	cluster, err := r.clusterLister.Get(clusterName)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	if cluster == nil {
```

Comment:
do we really need to check the namespace?

Reply:
removed